prov/util/src/util_srx.c

```diff
@@ -0,0 +1,1180 @@
+					"Error discarding message with peer\n");
+			}
+			ofi_buf_free(rx_entry);
+		}
+		((struct fi_context *)context)->internal[0] = rx_entry;
+	}
+
+	return ofi_cq_write(srx->cq, context, FI_TAGGED | FI_RECV,
+			    rx_entry->peer_entry.size, NULL, 0,
+			    rx_entry->peer_entry.tag);
+}
+
+ssize_t util_srx_generic_trecv(struct fid_ep *ep_fid, const struct iovec *iov,
+			       void **desc, size_t iov_count, fi_addr_t addr,
+			       void *context, uint64_t tag, uint64_t ignore,
+			       uint64_t flags)
+{
+	struct util_srx_ctx *srx;
+	struct util_rx_entry *rx_entry;
+	struct slist *queue;
+	ssize_t ret = FI_SUCCESS;
+
+	srx = container_of(ep_fid, struct util_srx_ctx, peer_srx.ep_fid);
+	assert(iov_count <= srx->iov_limit);
+	addr = srx->dir_recv ? addr : FI_ADDR_UNSPEC;
+
+	ofi_genlock_lock(&srx->lock);
+	if (flags & FI_PEEK) {
+		ret = util_srx_peek(srx, iov, desc, iov_count, addr,
+				    context, tag, ignore, flags);
+		goto out;
+	}
+
+	if (flags & FI_DISCARD) {
+		assert(flags & FI_CLAIM);
+		rx_entry = (struct util_rx_entry *)
+				(((struct fi_context *) context)->internal[0]);
+		ret = rx_entry->peer_entry.srx->peer_ops->discard_tag(
+							&rx_entry->peer_entry);
+		if (ret) {
+			FI_WARN(&core_prov, FI_LOG_EP_CTRL,
+				"Error discarding message with peer\n");
+		}
+		ret = ofi_cq_write(srx->cq, context, FI_TAGGED | FI_RECV,
+				    rx_entry->peer_entry.size, NULL, 0,
+				    rx_entry->peer_entry.tag);
+		ofi_buf_free(rx_entry);
+		goto out;
+	}
+
+	if (flags & FI_CLAIM) {
+		rx_entry = (struct util_rx_entry *)
+				(((struct fi_context *) context)->internal[0]);
+	} else {
+		rx_entry = util_search_unexp_tag(srx, addr, tag, ignore, true);
+		if (!rx_entry) {
+			queue = addr == FI_ADDR_UNSPEC ? &srx->tag_queue:
+				ofi_array_at(&srx->src_trecv_queues, addr);
+			assert(queue);
+			rx_entry = util_get_recv_entry(srx, iov, desc,
+						iov_count, addr, context, tag,
+						ignore, flags);
+			if (!rx_entry)
+				ret = -FI_ENOMEM;
+			else
+				slist_insert_tail((struct slist_entry *)
+						  (&rx_entry->peer_entry),
+						  queue);
+			goto out;
+		}
+	}
+	util_init_rx_entry(rx_entry, iov, desc, iov_count, addr, context, tag,
+			   flags);
+
+	srx->update_func(srx, rx_entry);
+	ret = rx_entry->peer_entry.srx->peer_ops->start_tag(
+					&rx_entry->peer_entry);
+out:
+	ofi_genlock_unlock(&srx->lock);
+	return ret;
+}
+
+ssize_t util_srx_generic_recv(struct fid_ep *ep_fid, const struct iovec *iov,
+			      void **desc, size_t iov_count, fi_addr_t addr,
+			      void *context, uint64_t flags)
+{
+	struct util_srx_ctx *srx;
+	struct util_rx_entry *rx_entry;
+	struct slist *queue;
+	ssize_t ret = FI_SUCCESS;
+
+	srx = container_of(ep_fid, struct util_srx_ctx, peer_srx.ep_fid);
+
+	if (flags & FI_MULTI_RECV)
+		return util_generic_mrecv(srx, iov, desc, iov_count, addr,
+					  context, flags);
+
+	assert(iov_count <= srx->iov_limit);
+	addr = srx->dir_recv ? addr : FI_ADDR_UNSPEC;
+
+	ofi_genlock_lock(&srx->lock);
```

Comment:
just find you add this lock back in util recvs... I need to remove them in efa's recv accordingly

Reply:
Yeah sorry should have mentioned it. Now all the owner srx functions have the lock. I think we will have to restrict users of the util srx to use the srx lock for the peer locking since they are so reliant on each other.

Reply:
>Now all the owner srx functions have the lock.

I think you still remove them in owner/peer ops like `get_msg`, which is good as they should be inside the progress engine (cq read) and it's owner's responsibility to grab that lock before driving peers' ep progress.

Reply:
I notice one caveat of having `srx->lock` inside util generic recv:

If the owner provider:
1. acquire a ep lock in its generic recv before it calls util recv
2. acquire a srx lock in its cq_read before calling it progress engine, which could acquire ep lock

There could be a deadlock in this situation, as thread A (generic recv) tries to acquire a srx lock held by thread B (cq read), while thread B tries to acquire a ep lock held by A.

I would have to either remove the ep lock in owner provider's generic recv, or move the srx lock from util recv to the owner's generic recv before acquiring ep lock.

Reply:
... Or we could move the srx lock inside ep lock in provider's ep_progress.

Reply:
@shijin-aws I think we need to add some lock definition to the srx lock and require it be the top most lock. shm (or the owner) might need the ep lock but that's really only needed because of the resource management in start_msg. So I think we can remove the shm acquire before the generic recv call and move that into the start_msg call so we always acquire in the same order (srx, ep). Does that seem reasonable to you?
I was chatting to Sean and he also suggested to passing in a genlock into srx for it to use as the srx lock which would allow the util srx code to use a lock that the owner already has when calling peer progress. Would that be of interest for efa?

Reply:
>I think we need to add some lock definition to the srx lock and require it be the top most lock ... So I think we can remove the shm acquire before the generic recv call and move that into the start_msg call so we always acquire in the same order (srx, ep). Does that seem reasonable to you?

Yes, I agree. I think we should make it clear that srx->lock should be the top most lock. In that case shm or efa's generic_recv does not require more lock before acquiring the srx lock.

> I was chatting to Sean and he also suggested to passing in a genlock into srx for it to use as the srx lock which would allow the util srx code to use a lock that the owner already has when calling peer progress. Would that be of interest for efa?

That may give us more flexibility, yes. But I am ok with the current implementation as long as we agree on the sequence of locking. I am happy to delegate the locking to util srx so efa does not need to have more code.

Reply:
@shijin-aws I've updated it to switch the locking so that shm acquires the ep lock in start_msg. Let me know if this still looks problematic.

Reply:
The change looks reasonable to me. I applied the same change in efa's start and discard to make them consistent. I ran the multi thread tests (osu_latency_mt) on both single node and multi node. Both look good to me.

Reply:
Excellent, thanks! Do you see any performance difference with these changes?

Reply:
osu_latency_mt does not show much difference.

Most of the application I ran do not request FI_THREAD_SAFE, so they shouldn't be impacted either.